client/rest-high-level/src/test/java/org/elasticsearch/client/dataframe/transforms/hlrc/DataFrameTransformStatsTests.java

```diff
@@ -58,48 +46,65 @@ public static NodeAttributes fromHlrc(org.elasticsearch.client.dataframe.transfo
             attributes.getAttributes());
     }
 
-    @Override
-    public org.elasticsearch.client.dataframe.transforms.DataFrameTransformState doHlrcParseInstance(XContentParser parser)
-            throws IOException {
-        return org.elasticsearch.client.dataframe.transforms.DataFrameTransformState.fromXContent(parser);
+    public static DataFrameTransformStats
+        fromHlrc(org.elasticsearch.client.dataframe.transforms.DataFrameTransformStats instance) {
+
+        return new DataFrameTransformStats(instance.getId(),
+            DataFrameTransformTaskState.fromString(instance.getTaskState().value()),
+            instance.getReason(),
+            fromHlrc(instance.getNode()),
+            DataFrameIndexerTransformStatsTests.fromHlrc(instance.getIndexerStats()),
+            DataFrameTransformCheckpointingInfoTests.fromHlrc(instance.getCheckpointingInfo()));
     }
 
     @Override
-    public DataFrameTransformState convertHlrcToInternal(org.elasticsearch.client.dataframe.transforms.DataFrameTransformState instance) {
-        return fromHlrc(instance);
+    public org.elasticsearch.client.dataframe.transforms.DataFrameTransformStats doHlrcParseInstance(XContentParser parser)
+            throws IOException {
+        return org.elasticsearch.client.dataframe.transforms.DataFrameTransformStats.fromXContent(parser);
     }
 
     @Override
-    protected DataFrameTransformState createTestInstance() {
-        return randomDataFrameTransformState();
+    public DataFrameTransformStats convertHlrcToInternal(
+        org.elasticsearch.client.dataframe.transforms.DataFrameTransformStats instance) {
+        return new DataFrameTransformStats(instance.getId(),
+                DataFrameTransformTaskState.fromString(instance.getTaskState().value()),
+                instance.getReason(),
+                fromHlrc(instance.getNode()),
+                DataFrameIndexerTransformStatsTests.fromHlrc(instance.getIndexerStats()),
+                DataFrameTransformCheckpointingInfoTests.fromHlrc(instance.getCheckpointingInfo()));
+    }
+
+    public static DataFrameTransformStats randomDataFrameTransformStats() {
+        return new DataFrameTransformStats(randomAlphaOfLength(10),
+            randomFrom(DataFrameTransformTaskState.values()),
+            randomBoolean() ? null : randomAlphaOfLength(100),
+            randomBoolean() ? null : randomNodeAttributes(),
+            // On the server side the stats has transform ID "_all" when embedded in another doc
```

Comment:
It is skipped when serializing.  The problem is that this test creates a server-side object, round-trips that through the HLRC back to another server-side object, then asserts equality.  After the round-tripping via serialization that discards the ID it ends up as `_all`.  Hence the original object has to be created with an ID of `_all` so that it is considered equal.

You're also correct that the cleanest solution with the way the rest of the code is today is just not to store a transform ID in the indexer stats.  At present when it's persisted to an index it's always inside a state-and-stats wrapper.  There is an idea to split the storage so that state and stats are persisted separately.  Obviously if that were done then we'd need an ID in the stats during persistence and this problem would come back.  But maybe it is best to leave the state and stats wrapped in a single document when they're persisted.  Grouping them may avoid consistency problems where the stats endpoint response includes old state and new stats (or vice-versa) for a particular transform due to the two docs becoming searchable at different times.  So in fact keeping state and stats grouped during persistence can solve two problems:

1. Guarantees consistency of state and indexer stats incorporated into any single stats endpoint response
2. Allows the complex/confusing code related to the sometimes-present ID in indexer stats to be deleted

Any thoughts @hendrikmuhs and @davidkyle?

Reply:
+1

Our discussions convinced me that we should stay with storing state and stats together. As explained above this avoids consistency problems and in addition we do not have to break backwards compatibility. The deletion of superfluous fields should not break BWC.